### PR TITLE
[feature] Testcontainers.Pulsar

### PR DESCRIPTION
**Pull request by eaba**
- [sdk net7.0] Testcontainers.Pulsar.Tests @HofmeisterAn. 
- SharpPulsar 2.11.0
- Akka.NET 1.5.0
- Apache Pulsar 2.11.0

### CODE REVIEW

File: Testcontainers.sln [hunk 1]
```diff
@@ -13,106 +13,107 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{673F23AE-769
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tests", "tests", "{7164F1FB-7F24-444A-ACD2-2C329C2B3CCF}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Testcontainers.Couchbase", "src\Testcontainers.Couchbase\Testcontainers.Couchbase.csproj", "{58E94721-2681-4D82-8D94-0B2F9DB0D575}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Testcontainers.Couchbase", "src\Testcontainers.Couchbase\Testcontainers.Couchbase.csproj", "{58E94721-2681-4D82-8D94-0B2F9DB0D575}"
```
Comment (HofmeisterAn) [inline Testcontainers.sln:16]: Can we please keep the old Visual Studio project type GUID for now and update to the new GUID in a separate PR? I would like to keep the changes for the module to a minimum. Please remove the unnecessary changes from the SLN.

File: global.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "6.0.400",
+    "version": "7.*",
```
Comment (HofmeisterAn) [inline global.json:3]: Please revert this. We stick to LTS versions.

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
+
+namespace Testcontainers.Pulsar
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs:2]:
```suggestion
namespace Testcontainers.Pulsar
```

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    /// <inheritdoc />
+    protected override PulsarConfiguration DockerResourceConfiguration { get; }
+
+    /// <inheritdoc />
+    public override PulsarContainer Build()
+    {
+      this.Validate();
+      return new PulsarContainer(this.DockerResourceConfiguration, TestcontainersSettings.Logger);
+    }
+    /// <inheritdoc />
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs:44]:
```suggestion
    }

    /// <inheritdoc />
```

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          .WithEnvironment("PULSAR_PREFIX_transactionCoordinatorEnabled", "true")
+          .WithEnvironment("PULSAR_PREFIX_brokerDeleteInactiveTopicsEnabled", "true")
+          .WithEnvironment("PULSAR_STANDALONE_USE_ZOOKEEPER", "1")
+          .WithEnvironment("PULSAR_PREFIX_exposingBrokerEntryMetadataToClientEnabled", "true")
+          .WithEnvironment("PULSAR_PREFIX_brokerEntryMetadataInterceptors", "org.apache.pulsar.common.intercept.AppendBrokerTimestampMetadataInterceptor,org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor")
+
+          .WithEntrypoint("/bin/sh", "-c")
+          .WithCommand("while [ ! -f " + StartupScriptFilePath + " ]; do sleep 0.1; done; " + StartupScriptFilePath)
+
+          .WithWaitStrategy(Wait.ForUnixContainer())
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs:64]: I think you need to add a proper wait strategy. Your wait strategy just waits until the container is running, but not until the service inside it is ready.

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+          {
+            const char lf = '\n';
+            var startupScript = new StringBuilder();
+            startupScript.Append("#!/bin/bash");
+            startupScript.Append(lf);
+            startupScript.Append("bin/apply-config-from-env.py conf/standalone.conf ");
+            startupScript.Append("&& bin/pulsar standalone --no-functions-worker ");
+            startupScript.Append("&& bin/pulsar initialize-transaction-coordinator-metadata -cs localhost:2181 -c standalone --initial-num-transaction-coordinators 2");
+
+            return container.CopyFileAsync(StartupScriptFilePath, Encoding.Default.GetBytes(startupScript.ToString()), 493, ct: ct);
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs:75]:
```suggestion
            return container.CopyFileAsync(StartupScriptFilePath, Encoding.Default.GetBytes(startupScript.ToString()), 493, ct: ct);
```

File: src/Testcontainers.Pulsar/PulsarConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
+
+namespace Testcontainers.Pulsar
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarConfiguration.cs:2]:
```suggestion
namespace Testcontainers.Pulsar
```

File: src/Testcontainers.Pulsar/PulsarConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  [PublicAPI]
+  public sealed class PulsarConfiguration : ContainerConfiguration
+  {
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PulsarConfiguration" /> class.
+    /// </summary>
+    public PulsarConfiguration()
+    {
+    }
+    /// <summary>
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarConfiguration.cs:14]:
```suggestion
    }

    /// <summary>
```

File: src/Testcontainers.Pulsar/PulsarContainer.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+
+namespace Testcontainers.Pulsar
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarContainer.cs:2]:
```suggestion
namespace Testcontainers.Pulsar
```

File: src/Testcontainers.Pulsar/Testcontainers.Pulsar.csproj [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="JetBrains.Annotations" Version="2022.3.1"/>
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="$(SolutionDir)src/Testcontainers/Testcontainers.csproj"/>
+  </ItemGroup>
+</Project>
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/Testcontainers.Pulsar.csproj:13]:
```suggestion
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
        <LangVersion>latest</LangVersion>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="JetBrains.Annotations" Version="2022.3.1"/>
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="$(SolutionDir)src/Testcontainers/Testcontainers.csproj"/>
    </ItemGroup>
</Project>
```

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+{
+  /// <inheritdoc cref="ContainerBuilder{TBuilderEntity, TContainerEntity, TConfigurationEntity}" />
+  [PublicAPI]
+  public sealed class PulsarBuilder : ContainerBuilder<PulsarBuilder, PulsarContainer, PulsarConfiguration>
+  {
+    public const string PulsarImage = "apachepulsar/pulsar-all:2.11.0";
+
+    public ushort PulsarPort = 6650;
+    public ushort PulsarAdminPort = 8080;
+    public ushort PulsarSQLPort = 8081;
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs:12]:
```suggestion
    public const ushort PulsarPort = 6650;
    public const ushort PulsarAdminPort = 8080;
    public const ushort PulsarSQLPort = 8081;
```

File: src/Testcontainers.Pulsar/PulsarBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      return new PulsarContainer(this.DockerResourceConfiguration, TestcontainersSettings.Logger);
+    }
+    /// <inheritdoc />
+    protected override PulsarBuilder Init()
+    {
+      return base.Init()
+          .WithImage(PulsarImage)
+          .WithPortBinding(this.PulsarPort, this.PulsarPort)
+          .WithPortBinding(this.PulsarAdminPort, this.PulsarAdminPort)
+          .WithPortBinding(this.PulsarSQLPort, this.PulsarSQLPort)
```
Comment (HofmeisterAn) [inline src/Testcontainers.Pulsar/PulsarBuilder.cs]:
```suggestion
          .WithPortBinding(PulsarPort, true)
          .WithPortBinding(PulsarAdminPort, true)
          .WithPortBinding(PulsarSQLPort, true)
```

We never bind static host ports, please use random assigned host ports instead. Your container can resolve the port with `GetMappedPublicPort(ushort)`, see:

- https://dotnet.testcontainers.org/api/best_practices/ (2. and 3.)
- https://dotnet.testcontainers.org/api/create_docker_container/#examples

File: tests/Testcontainers.Pulsar.Tests/.editorconfig [hunk 13]
```diff
@@ -0,0 +1 @@
+root = true
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Pulsar.Tests/.editorconfig:1]: Please add this file to the src project too.

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Thanks for your PR. Your efforts are highly appreciated, most parts look good and merely require cosmetic changes.

However, before I can proceed with reviewing your PR, I have a few change requests that need to be addressed. Specifically, I need you to remove the static host port binding and remove any changes that are not relevant to this PR.

Once you make these updates, I will be happy to continue with the review process. Thank you again for your contribution, and I look forward to seeing the updated changes.

File: global.json (hunk 2 above)
Comment (eaba) [reply global.json:3]: `SharpPulsar` is `NET7.0` only

File: src/Testcontainers.Pulsar/PulsarBuilder.cs (hunk 5 above)
Comment (eaba) [reply src/Testcontainers.Pulsar/PulsarBuilder.cs:64]: I followed `Testcontainers.Kafka`

File: global.json (hunk 2 above)
Comment (HofmeisterAn) [reply global.json:3]: Why cannot we use the `DotPulsar`? It supports .NET Standard 2.0 and above. It should be enough to test the container configuration and does not require so many changes.

File: src/Testcontainers.Pulsar/PulsarBuilder.cs (hunk 5 above)
Comment (HofmeisterAn) [reply src/Testcontainers.Pulsar/PulsarBuilder.cs:64]: Kafka waits for a log message, not just until the container is running:

https://github.com/testcontainers/testcontainers-dotnet/blob/ebedc9a5d23566c8353cff18c8320ffd2b6b9b58/src/Testcontainers.Kafka/KafkaBuilder.cs#L67

**Pull request closed without merging**